File: Swift/ExampleTests/common/AddressComponentsString.swift
public struct AddressComponentsString: CustomStringConvertible, Equatable {
    public let zip: String
    public let streetName: String
    public let city: String
    public let preDirection: String
    public let suffixDirection: String
    public let state: String
    public let suffixType: String

    public init(zip: String, streetName: String, city: String, preDirection: String, suffixDirection: String, state: String, suffixType: String) {
        self.zip = zip
        self.streetName = streetName
        self.city = city
        self.preDirection = preDirection
        self.suffixDirection = suffixDirection
        self.state = state
        self.suffixType = suffixType
    }

    public init(fromArray v: [String]) {
        self.zip = v.count > 0 ? v[0] : ""
        self.streetName = v.count > 1 ? v[1] : ""
        self.city = v.count > 2 ? v[2] : ""
        self.preDirection = v.count > 3 ? v[3] : ""
        self.suffixDirection = v.count > 4 ? v[4] : ""
        self.state = v.count > 5 ? v[5] : ""
        self.suffixType = v.count > 6 ? v[6] : ""
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    public static func fromText(_ text: String) -> AddressComponentsString {
        let parts = Tokens.require(text, 7, "AddressComponents")
        return AddressComponentsString(zip: parts[0], streetName: parts[1], city: parts[2], preDirection: parts[3], suffixDirection: parts[4], state: parts[5], suffixType: parts[6])
    }

    public var description: String {
        return Tokens.token(zip) + " " + Tokens.token(streetName) + " " + Tokens.token(city) + " " + Tokens.token(preDirection) + " " + Tokens.token(suffixDirection) + " " + Tokens.token(state) + " " + Tokens.token(suffixType)
    }

    public static let dncString = "?DNC?"

    public static func == (a: AddressComponentsString, b: AddressComponentsString) -> Bool {
        return (a.zip == dncString || b.zip == dncString || a.zip == b.zip)
            && (a.streetName == dncString || b.streetName == dncString || a.streetName == b.streetName)
            && (a.city == dncString || b.city == dncString || a.city == b.city)
            && (a.preDirection == dncString || b.preDirection == dncString || a.preDirection == b.preDirection)
            && (a.suffixDirection == dncString || b.suffixDirection == dncString || a.suffixDirection == b.suffixDirection)
            && (a.state == dncString || b.state == dncString || a.state == b.state)
            && (a.suffixType == dncString || b.suffixType == dncString || a.suffixType == b.suffixType)
    }
}

File: Python/ExampleTests/common/address_components_string.py
from . import tokens
DNC_STRING = '?DNC?'


class AddressComponentsString:
    def __init__(self, zip: str = '', street_name: str = '', city: str = '', pre_direction: str = '', suffix_direction: str = '', state: str = '', suffix_type: str = ''):
        self.zip = zip
        self.street_name = street_name
        self.city = city
        self.pre_direction = pre_direction
        self.suffix_direction = suffix_direction
        self.state = state
        self.suffix_type = suffix_type

    @classmethod
    def from_list(cls, values):
        v = list(values)
        return cls(
            v[0] if len(v) > 0 else '',
            v[1] if len(v) > 1 else '',
            v[2] if len(v) > 2 else '',
            v[3] if len(v) > 3 else '',
            v[4] if len(v) > 4 else '',
            v[5] if len(v) > 5 else '',
            v[6] if len(v) > 6 else ''
        )

    @classmethod
    def from_text(cls, text):
        """Builds from the text form, e.g. Money as '25 USD'."""
        parts = tokens.require(text, 7, 'AddressComponents')
        return cls(
            parts[0],
            parts[1],
            parts[2],
            parts[3],
            parts[4],
            parts[5],
            parts[6]
        )

    def __str__(self):
        return (tokens.token(self.zip) + ' ' +
                tokens.token(self.street_name) + ' ' +
                tokens.token(self.city) + ' ' +
                tokens.token(self.pre_direction) + ' ' +
                tokens.token(self.suffix_direction) + ' ' +
                tokens.token(self.state) + ' ' +
                tokens.token(self.suffix_type))

    def _key(self):
        return (self.zip, self.street_name, self.city, self.pre_direction, self.suffix_direction, self.state, self.suffix_type)

    def __eq__(self, other):
        if not isinstance(other, AddressComponentsString):
            return NotImplemented
        return all(a == b or a == DNC_STRING or b == DNC_STRING
                   for a, b in zip(self._key(), other._key()))

    def __hash__(self):
        return hash(self._key())

    def __repr__(self):
        return f'AddressComponentsString({self})'
